ts/src/connections/tunnelRelayTunnelClient.ts

```diff
@@ -499,15 +499,15 @@ export class TunnelRelayTunnelClient extends TunnelConnectionSession implements
         return this.sshSession?.getService(PortForwardingService) ?? undefined;
     }
 
-    public async refreshPorts(): Promise<void> {
+    public async refreshPorts(cancellation?: CancellationToken): Promise<void> {
         if (!this.sshSession || this.sshSession.isClosed) {
             throw new Error('Not connected.');
         }
 
         const request = new SessionRequestMessage();
         request.requestType = 'RefreshPorts';
         request.wantReply = true;
-        await this.sshSession.request(request);
+        await this.sshSession.request(request, cancellation);
```

Comment:
How long would a request take to get successfully canceled? I'm just wondering if we go with this approach, how long would the user have to wait before retrying.

Reply:
Cancelling an SSH session request causes the SSH session to close, because it can interrupt sending/receiving protocol messages and leave the protocol in an unknown state. The cancellation parameter for the `SshSession.request()` method is meant to be used only when shutting down the session. So you can't cancel and retry a session request like this.

I don't think there is much value in retrying the request from here anyway: the SSH protocol guarantees the message will be received by the other side (unless the entire session fails). The right place to add retry would be on the [host's handling of that request](https://github.com/microsoft/dev-tunnels/blob/869adcec1c69b73a011a407a7ea8143e9ad54f79/ts/src/connections/tunnelRelayTunnelHost.ts#L632), when it calls the service to request updated ports info.

Reply:
What would mean to cancel the `refreshPorts` request from the host side (which ultimately calls `refreshTunnel`)? Would that just cancel the service request and the SSH session will remain intact? Or would that require the client to re-establish the SSH connection?

Reply:
Yes on the host side cancelling would just cancel the service request and leave any client SSH sessions intact. (After the response from the service it sends notification messages about any changes over SSH to connected clients, but that part can't be cancelled.)

Reply:
thanks! Will add retry logic if we see that it is necessary after the next round of results.